Some changes on top of 0.9

fbeister opened the pull request:
These are the changes we made on top of 0.9 in order to prevent a memory leak and unnecessary reconnects

andsel commented:
Sorry but I can't understand where your changes solves which leak problems

fbeister commented:
Now I have bit more time to elaborate:

* Many messages are stored in the MessagesStore and never deleted. They would have been deleted on reconnect but for our usecase that never really happens and thus the MessageStore grows infinitely. We pulled out the remove method to the interface and added calls to it in relevant places. As we remove the message, there is no way to call the onAcknowledged handler anymore (and we don't need it so we just removed it, sorry :-D )

* Using isWritable means that we completely skip the Netty queue. isWritable returns false if the channel is active but currently busy - enqueuing in that situation is IMHO the wrong decision.

* We removed the PUBACK handler as a broker shouldn't receive a PUBACK (or should it? we are mere beginners in working with MQTT so please tell me if we are wrong here!)

* Messages with QOS2 were stored with two different GUIDs, this is fixed in MessagesPublisher


I'm keen to get your feedback on our changes as you are the expert here. Sorry for not rebasing this onto the latest master but this is happening during paid working hours :-D

andsel commented:
On Wed, Mar 22, 2017 at 10:41 PM, fbeister <notifications@github.com> wrote:

> Now I have bit more time to elaborate:
>

This is a fail in the protocol spec, if a client is with clean session
false and never reconnect, the broker must store the messages. We could
think to something like TTL or a command line too that clean old messages.
In this are MQTT v5 solve this issue introducing the concept of TTL.

>
>    -
>
>    Many messages are stored in the MessagesStore and never deleted. They
>    would have been deleted on reconnect but for our usecase that never really
>    happens and thus the MessageStore grows infinitely. We pulled out the
>    remove method to the interface and added calls to it in relevant places. As
>    we remove the message, there is no way to call the onAcknowledged handler
>    anymore (and we don't need it so we just removed it, sorry :-D )
>
> You are right, I'm the way to move the persistent queue out of MapDB so
that impact also this queuing, I expect to change this.

>
>
>    -
>
>    Using isWritable means that we completely skip the Netty queue.
>    isWritable returns false if the channel is active but currently busy -
>    enqueuing in that situation is IMHO the wrong decision.
>
> No, when the broker publish QoS1 to a Subscriber it has to listen for the
PUBACK from the subscriber (the same for the QoS2 messages, the publish
from the server is dual to what append when the broker receives a publish)

>
>
>    -
>
>    We removed the PUBACK handler as a broker shouldn't receive a PUBACK
>    (or should it? we are mere beginners in working with MQTT so please tell me
>    if we are wrong here!)
>
> I've to check this, is not good if the broker has such behaviour

>
>    -
>
>    Messages with QOS2 were stored with two different GUIDs, this is fixed
>    in MessagesPublisher
>
> Can you share more details about the project (just to let me know in which
contexts Moquette are used)

> I'm keen to get your feedback on our changes as you are the expert here.
> Sorry for not rebasing this onto the latest master but this is happening
> during paid working hours :-D
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/andsel/moquette/pull/287#issuecomment-288548890>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_Y0WPTVt3yTKCMQg04scftbzkQPHKvks5roZWVgaJpZM4Mj1HN>
> .
>

fbeister commented:
On 22.03.2017 23:08, Andrea Selva wrote:
> On Wed, Mar 22, 2017 at 10:41 PM, fbeister <notifications@github.com> wrote:
> 
>> Now I have bit more time to elaborate:
>>
> 
> This is a fail in the protocol spec, if a client is with clean session
> false and never reconnect, the broker must store the messages. We could
> think to something like TTL or a command line too that clean old messages.
> In this are MQTT v5 solve this issue introducing the concept of TTL.

Can't we delete the message when all clients have acknowledged the
message (immediately for Qos0, on PUBACK for QoS1 and on PUBCOMP for
Qos2) as done in the pull request? Moquette is keeping messages it has
already received acknowledgements for.

> 
>>
>> -
>>
>> Many messages are stored in the MessagesStore and never deleted. They
>> would have been deleted on reconnect but for our usecase that never really
>> happens and thus the MessageStore grows infinitely. We pulled out the
>> remove method to the interface and added calls to it in relevant
> places. As
>> we remove the message, there is no way to call the onAcknowledged handler
>> anymore (and we don't need it so we just removed it, sorry :-D )
>>
>> You are right, I'm the way to move the persistent queue out of MapDB so
> that impact also this queuing, I expect to change this.
> 
>>
>>
>> -
>>
>> Using isWritable means that we completely skip the Netty queue.
>> isWritable returns false if the channel is active but currently busy -
>> enqueuing in that situation is IMHO the wrong decision.
>>
>> No, when the broker publish QoS1 to a Subscriber it has to listen for the
> PUBACK from the subscriber (the same for the QoS2 messages, the publish
> from the server is dual to what append when the broker receives a publish)
> 

Oh, I see. I have put the respective call to processPubAck back in. The
onAcknowledge handler doesn't work after making the changes to the
message deletion so in our codebase we just removed it. Depending on the
outcome of the above discussion, the handling here might have to be
revamped.

>>
>>
>> -
>>
>> We removed the PUBACK handler as a broker shouldn't receive a PUBACK
>> (or should it? we are mere beginners in working with MQTT so please
> tell me
>> if we are wrong here!)
>>
>> I've to check this, is not good if the broker has such behaviour
> 
>>
>> -
>>
>> Messages with QOS2 were stored with two different GUIDs, this is fixed
>> in MessagesPublisher
>>
>> Can you share more details about the project (just to let me know in which
> contexts Moquette are used)

We do detection of UAV (commercial drones) with distributed sensors.

> 
>> I'm keen to get your feedback on our changes as you are the expert here.
>> Sorry for not rebasing this onto the latest master but this is happening
>> during paid working hours :-D
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/andsel/moquette/pull/287#issuecomment-288548890>,
> or mute
>> the thread
>>
> <https://github.com/notifications/unsubscribe-auth/AA_Y0WPTVt3yTKCMQg04scftbzkQPHKvks5roZWVgaJpZM4Mj1HN>
>> .
>>
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/andsel/moquette/pull/287#issuecomment-288555314>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHLHvjzLxdjQLqI5y2nOZwBEqmH0kmXxks5roZv5gaJpZM4Mj1HN>.
> 


-- 
Frederic Beister
PGP Fingerprint = 75F7 F0DF 7DCA E46D EBB6 679C CA38 3D52 A3F5 478B
